asn1c/asn1c.1
.TH asn1c 1 "2014-10-23" "ASN.1 Compiler"
.if n .ad l \" Turn off justifications
.nh \" Turn off hyphenation; can damage formatting in technical documents.
.SH NAME
asn1c \(em the ASN.1 Compiler
.SH SYNOPSIS
asn1c [\fB\-E\fR [\fB-F\fR] | \fB\-P\fR | \fB\-R\fR]
      [\fB\-S\fR\fIdir\fR] [\fB-X\fR]
      [\fB\-W\fR\fIdebug-\fR...] [\fB\-f\fR\fIoption\fR] [\fB\-gen-\fR\fIoption\fR] [\fB\-pdu=\fR{\fBall\fR|\fBauto\fR|\fIType\fR}\fR]
      [\fB\-print-\fR\fIoption\fR]
      \fIinput-filenames\fR...
.SH DESCRIPTION
asn1c compiles ASN.1 specifications into a set of
target language (C/C++) encoders and decoders for BER, DER, PER, XER
and other encoding rules.
.SH OPTIONS
.SS Stage Selection Options
.TP 4
.B \-E
Run the parsing stage only.
Print the reconstructed ASN.1 text.
.TP 4
.B \-F
Used together with
.B \-E\c
, instructs the compiler to stop after the ASN.1 syntax
tree fixing stage and dump the reconstructed ASN.1 specification
to the standard output.
.TP 4
.B \-P
Dump the compiled output to the standard output instead of creating the
target language files on disk.
.TP 4
.B \-R
Restrict the compiler to generate only the ASN.1 tables,
omitting the usual support code.
.TP 4
\fB\-S\fR \fIdirectory\fR
Use the specified directory with ASN.1 skeleton files.
.TP 4
.B \-X
Generate an XML DTD schema for the specified ASN.1 files.
.SS Warning Options
.TP 4
.B \-Werror
Treat warnings as errors; abort if any warning is produced.
.TP 4
.B \-Wdebug-lexer
Enable lexer debugging during the ASN.1 parsing stage.
.TP 4
.B \-Wdebug-fixer
Enable ASN.1 syntax tree fixer debugging during the fixing stage.
.TP 4
.B \-Wdebug-compiler
Enable debugging during the actual compile time.
.SS Language Options
.TP 4
.B \-fbless-SIZE
Allow SIZE() constraint for INTEGER, ENUMERATED, and other types for which this
constraint is normally prohibited by the standard.
This is a violation of ASN.1 standard, and the compiler may
fail to\ produce a meaningful code.
.TP 4
.B \-fcompound-names
Using this option prevents name collisions in the target source code
by using complex names for target language structures.
(Name\ collisions may occur if the ASN.1 module reuses the same identifiers
in multiple contexts).
.TP 4
.B \-findirect-choice
When generating code for a CHOICE type, compile the CHOICE members as indirect
pointers instead of declaring them inline.
Consider using this option together with
.B \-fno-include-deps
to prevent circular references.
.TP 4
.B \-fincludes-quoted
Generate #include lines in "double" instead of <angle> quotes.
.TP 4
.BI "\-fknown-extern-type="<name>
Pretend the specified type is known.
The compiler will assume the target language source files
for the given type have been provided manually.
.TP 4
.B \-fline-refs
Include ASN.1 module's line numbers in generated code comments.
.TP 4
.B \-fno-constraints
Do not generate ASN.1 subtype constraint checking code.
This may make a shorter executable.
.TP 4
.B \-fno-include-deps
Do not generate courtesy #include lines for non-critical type dependencies.
Helps prevent namespace collisions.
.TP 4
.B \-funnamed-unions
Enable unnamed unions in the definitions of target language's structures.
.TP 4
.B \-fwide-types
Use the unbounded size data types (INTEGER_t, ENUMERATED_t, REAL_t) by default,
instead of the native machine's data types (long, double).
.SS Codecs Generation Options
.TP 4
.B \-gen-PER
Generate the Packed Encoding Rules (PER) support code.
.TP 4
.B \-pdu=\fR{\fBall\fR|\fBauto\fR|\fIType\fR}
Create a PDU table for specified types, or discover Protocol Data Units automatically.
In case of -pdu=\fBall\fR, all ASN.1 types defined in all modules will form a PDU table.
In case of -pdu=\fBauto\fR, all types not referenced by any other type will form a PDU table.
If\ \fIType\fR is an ASN.1 type identifier, it is added to a PDU table.
The\ last form may be specified multiple times to add any number of PDUs.
.SS Output Options
.TP 4
.B \-print-constraints
When -EF are also specified, this option forces the compiler to explain
its internal understanding of subtype constraints.
.TP 4
.B \-print-lines
Generate "-- #line" comments in \fB-E\fR output.
.SH SEE ALSO
.TP 4
\&\fIunber\fR\|(1), \&\fIenber\fR\|(1)
.SH AUTHORS
Lev Walkin <vlm@lionet.info>


ChangeLog

0.9.??:
    * Fixed generated file permissions introduced by earlier pull (issue/32).
    * Constness patch #46 by Wim L <wiml@omnigroup.com> (41bbf1c..78d604f).
    * Win32/MinGW patch #31 by Frank Morgner (frankmorgner).

0.9.27: 2014-Sep-17
    * Build structure fixes for portability.
    * Disable long debug of ASN.1 operations during `make check` stage.
    * Introduce --enable-ASN_DEBUG configure flag to emit long debug output.

0.9.26: 2014-Sep-11
    * Updated asn1c-usage.pdf.
    * Made -fskeletons-copy a default option; removed cmdline option.
    * Made -fnative-types a default option; removed cmdline option.
      Added -fwide-types for backwards compatibility.
    * Add -fline-refs to avoid being unfriendly to version control systems.
    * Default constraint checking fix.
      Thanks to Bartosz Marcinkiewicz <bma@megawatt.com.pl>
    * Get rid of non-standard pointer arithmetics on void* pointer.
    * PER-encoding of integers wider than 32 bits.

0.9.24:    2013-Mar-16

    * GeneralizedTime fix for working with time offsets which are not
      representable in whole hours. (Severity: low; Security impact: low)
      Thanks to IP Fabrics, Inc.
    * Added -fincludes-quoted to asn1c to generate #includes in "double"
      instead of <angle> quotes.
    * PER encoding correctness fix. (Severity: high; Security impact: low)
      Reported by Grzegorz Aksamit.
    * ENUMERATED extension values check relaxed. Reported by Gabriel Burca.
    * Fixed decimal decoding of REAL values in -fnative-types mode
      (Severity: medium; Security impact: medium)
    * Removed webcgi subproject (introduced in 2004) as useless.

0.9.22:    2008-Nov-19

    * Added -pdu=all and -pdu=<type> switches to asn1c.
    * Added PER support for most known-multiplier string types:
      IA5String, VisibleString, PrintableString, BMPString, UniversalString;
      useful types: GeneralizedTime, UTCTime, ObjectDescriptor;
      as well as REAL and OBJECT IDENTIFIER.
      TODO: SET type.
    * Added PER support for extensions in SEQUENCE.
    * Multiple enhancements by Daniele Varrazzo <daniele.varrazzo@gmail.com>
    * Fixed explicit tagging of an in-lined constructed type.
      (Severity: low; Security impact: none).
    * Memory leak fixed in SET OF / SEQUENCE OF when wrong encoding
      is encountered. (Severity: medium; Security impact: low)
    * Fixed extensibility handling of second SEQUENCE production.
      (Severity: low; Security impact: none)
    * Added DEFAULT handling for known multiplier string.
    * Added a sample OMA ULP decoder (./examples/sample.source.ULP).
    * Added full-width 32-bit integer encoding support in PER.
    * Fixed 1-byte OOB write issue with non-standard and Windows
      memory allocators (Severity: low; Security impact: medium).
      Reported by Sheng Yu.

0.9.21:    2006-Sep-17

    * skeletons/standard-modules directory is now used for standard types.
    * Fixed class field access problem (Test case 98)
      (Severity: medium; Security impact: none)
    * Refactored Information Object Classes parsing.
    * Refactored Parameterization support.
    * [typedef enum foo {}] foo_e; is now e_foo, see #1287989
    * Refactored ValueSetTypeAssignment parsing.
    * asn-decoder-template.c renamed into converter-sample.c
    * MEGACO (Media Gateway Control Protocol) decoder sample added.
    * First release of PER encoder (does not encode SETs yet).
    * Added a sample LDAP decoder (./examples/sample.source.LDAP3).
    * Added a sample 3GPP RRC decoder (./examples/sample.source.RRC).
    * Added a sample MEGACO (H.248.1) decoder (sample.source.MEGACO).

0.9.20:    2006-Mar-06

    * SET OF CHOICE, SEQUENCE OF CHOICE and a certain named S/O types
      are represented differently in XER. THIS IS AN INCOMPATIBLE CHANGE.
      (Test case 70) (Severity: low; Security impact: low)
    * asn1c: Removed -ftypes88 command line option.
    * Started PER implementation. Somewhat experimental!

0.9.19:    2005-Oct-06

    * A proper solution to circular references. No kludge flags
      should be necessary anymore to produce reference-free code:
      recursive dependencies are resolved automatically.
    * Test cases 73 & 92 keep track of various circular references.
    * Introduced compiler directives to allow finer control over the
      generated code ("--<ASN1C...>--" in comments), (Test case 93).
    * New feature for unber(1): -s <skip> bytes.
    * Mandatory elements map for SET was not getting generated properly.
      (Test case 94) (Severity: high; Security impact: low)
    * asn1c: new command line option: -fskeletons-copy.

0.9.18:    2005-Aug-14

    * The obsolete X.208 syntax is handled gracefully now (compound types'
      member names are invented on the fly). (Test case 87).
    * Generating enumeration tables for INTEGER types (Test case 88).
    * Generating enumeration tables for BIT STRING types (Test case 89).
    * Conditional INTEGER/ENUMERATED representation: long vs. INTEGER_t
      type is chosen based on PER visible constraints (Test cases 90, 91).
    * Union structures for CHOICE type are not anonymous anymore.
    * Made unber(1) program more verbose: effective structure length is
      now displayed as L="<n>" at the closing tag. (Test check-xxber.sh).
    * Fixed unber(1)'s -1 switch for indefinite encoding length.
    * New command line option for unber(1): -m to enable minimalistic mode.

0.9.17:    2005-Aug-07

    * Tagging mode is implicitly IMPLICIT if AUTOMATIC TAGS is used, #30.6.
      (Test cases 55, 86).
    * Started to use extern "C" {} in skeleton headers.
    * Introduced -finline-choice command line option to fix
      yet another class of circular references.

0.9.16:    2005-July-04

    * GeneralizedTime API now supports fractions of seconds.
      Thanks to Bent Nicolaisen <BN@JAI.com> for support.
    * ASN.1 parser has been tweaked to allow parsing something like

asn1c/asn1c.c
/*-
 * Copyright (c) 2003-2014
 * 	Lev Walkin <vlm@lionet.info>. All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 *
 * $Id$
 */
/*
 * This is the program that connects the libasn1* libraries together.
 * It uses them in turn to parse, fix and then compile or print the ASN.1 tree.
 */
#include "sys-common.h"

#undef  COPYRIGHT
#define COPYRIGHT       \
	"Copyright (c) 2003-2014 Lev Walkin <vlm@lionet.info>\n"

#include <asn1parser.h>		/* Parse the ASN.1 file and build a tree */
#include <asn1fix.h>		/* Fix the ASN.1 tree */
#include <asn1print.h>		/* Print the ASN.1 tree */
#include <asn1compiler.h>	/* Compile the ASN.1 tree */

#include <asn1c_compat.h>	/* Portable basename(3) and dirname(3) */

#ifdef	_WIN32
#include <io.h>
#include <direct.h>
#else
#include <dirent.h>
#endif

static void usage(const char *av0);	/* Print the Usage screen and exit */
static int importStandardModules(asn1p_t *asn, const char *skeletons_dir);

int
main(int ac, char **av) {
	enum asn1p_flags     asn1_parser_flags	= A1P_NOFLAGS;
	enum asn1f_flags     asn1_fixer_flags	= A1F_NOFLAGS;
	enum asn1c_flags     asn1_compiler_flags= A1C_NO_C99;
	enum asn1print_flags asn1_printer_flags	= APF_NOFLAGS;
	int print_arg__print_out = 0;	/* Don't compile, just print parsed */
	int print_arg__fix_n_print = 0;	/* Fix and print */
	int warnings_as_errors = 0;	/* Treat warnings as errors */
	char *skeletons_dir = NULL;	/* Directory with supplementary stuff */
	asn1p_t *asn = 0;		/* An ASN.1 parsed tree */
	int ret;			/* Return value from misc functions */
	int ch;				/* Command line character */
	int i;				/* Index in some loops */

	/*
	 * Process command-line options.
	 */
	while((ch = getopt(ac, av, "EFf:g:hLPp:RS:vW:X")) != -1)
	switch(ch) {
	case 'E':
		print_arg__print_out = 1;
		break;
	case 'F':
		print_arg__fix_n_print = 1;
		break;
	case 'f':
		if(strcmp(optarg, "all-defs-global") == 0) {
			asn1_compiler_flags |= A1C_ALL_DEFS_GLOBAL;
		} else if(strcmp(optarg, "bless-SIZE") == 0) {
			asn1_fixer_flags |= A1F_EXTENDED_SizeConstraint;
		} else if(strcmp(optarg, "compound-names") == 0) {
			asn1_compiler_flags |= A1C_COMPOUND_NAMES;
		} else if(strcmp(optarg, "indirect-choice") == 0) {
			asn1_compiler_flags |= A1C_INDIRECT_CHOICE;
		} else if(strncmp(optarg, "known-extern-type=", 18) == 0) {
			char *known_type = optarg + 18;
			ret = asn1f_make_known_external_type(known_type);
			assert(ret == 0 || errno == EEXIST);
		} else if(strcmp(optarg, "native-types") == 0) {
			fprintf(stderr, "-f%s: Deprecated option\n", optarg);
			asn1_compiler_flags &= ~A1C_USE_WIDE_TYPES;
		} else if(strcmp(optarg, "wide-types") == 0) {
			asn1_compiler_flags |= A1C_USE_WIDE_TYPES;
		} else if(strcmp(optarg, "line-refs") == 0) {
			asn1_compiler_flags |= A1C_LINE_REFS;
		} else if(strcmp(optarg, "no-constraints") == 0) {
			asn1_compiler_flags |= A1C_NO_CONSTRAINTS;
		} else if(strcmp(optarg, "no-include-deps") == 0) {
			asn1_compiler_flags |= A1C_NO_INCLUDE_DEPS;
		} else if(strcmp(optarg, "includes-quoted") == 0) {
			asn1_compiler_flags |= A1C_INCLUDES_QUOTED;
		} else if(strcmp(optarg, "unnamed-unions") == 0) {
			asn1_compiler_flags |= A1C_UNNAMED_UNIONS;
		} else if(strcmp(optarg, "skeletons-copy") == 0) {
			fprintf(stderr, "-f%s: Deprecated option\n", optarg);
			asn1_compiler_flags &= ~A1C_LINK_SKELETONS;
		} else if(strcmp(optarg, "link-skeletons") == 0) {
			asn1_compiler_flags |= A1C_LINK_SKELETONS;
		} else {
			fprintf(stderr, "-f%s: Invalid argument\n", optarg);
			exit(EX_USAGE);
		}
		break;
	case 'g':
		if(strcmp(optarg, "en-PER") == 0) {
			asn1_compiler_flags |= A1C_GEN_PER;
		} else {
			fprintf(stderr, "-g%s: Invalid argument\n", optarg);
			exit(EX_USAGE);
		}
		break;
	case 'h':
		usage(av[0]);
	case 'P':
		asn1_compiler_flags |= A1C_PRINT_COMPILED;
		asn1_compiler_flags &= ~A1C_NO_C99;
		break;
	case 'p':
		if(strncmp(optarg, "du=", 3) == 0) {
			char *pduname = optarg + 3;
			if(strcmp(pduname, "all") == 0) {
				asn1_compiler_flags |= A1C_PDU_ALL;
			} else if(strcmp(pduname, "auto") == 0) {
				asn1_compiler_flags |= A1C_PDU_AUTO;
			} else if(pduname[0] >= 'A' && pduname[0] <= 'Z') {
				asn1c__add_pdu_type(pduname);
				asn1_compiler_flags |= A1C_PDU_TYPE;
			} else {
				fprintf(stderr, "-pdu=%s"
					": expected -pdu={all|auto|Type}\n",
					pduname);
				exit(EX_USAGE);
			}
		} else if(strcmp(optarg, "rint-class-matrix") == 0) {
			asn1_printer_flags |= APF_PRINT_CLASS_MATRIX;
		} else if(strcmp(optarg, "rint-constraints") == 0) {
			asn1_printer_flags |= APF_PRINT_CONSTRAINTS;
		} else if(strcmp(optarg, "rint-lines") == 0) {
			asn1_printer_flags |= APF_LINE_COMMENTS;
		} else {
			fprintf(stderr, "-p%s: Invalid argument\n", optarg);
			exit(EX_USAGE);
		}
		break;
	case 'R':
		asn1_compiler_flags |= A1C_OMIT_SUPPORT_CODE;
		break;
	case 'S':
		skeletons_dir = optarg;
		break;
	case 'v':
		fprintf(stderr, "ASN.1 Compiler, v" VERSION "\n" COPYRIGHT);
		exit(0);
		break;
	case 'W':
		if(strcmp(optarg, "error") == 0) {
			warnings_as_errors = 1;
			break;
		} else if(strcmp(optarg, "debug-lexer") == 0) {
			asn1_parser_flags |= A1P_LEXER_DEBUG;
			break;
		} else if(strcmp(optarg, "debug-fixer") == 0) {
			asn1_fixer_flags |= A1F_DEBUG;
			break;
		} else if(strcmp(optarg, "debug-compiler") == 0) {
			asn1_compiler_flags |= A1C_DEBUG;
			break;
		} else {
			fprintf(stderr, "-W%s: Invalid argument\n", optarg);
			exit(EX_USAGE);
		}
		break;
	case 'X':
		print_arg__print_out = 1;	/* Implicit -E */
		print_arg__fix_n_print = 1;	/* Implicit -F */
		asn1_printer_flags |= APF_PRINT_XML_DTD;
		break;
	default:
		usage(av[0]);
	}

	/*
	 * Validate the options combination.
	 */
	if(!print_arg__print_out) {
		if(print_arg__fix_n_print) {
			fprintf(stderr, "Error: -F requires -E\n");
			exit(EX_USAGE);
		}
		if(asn1_printer_flags) {
			fprintf(stderr, "Error: "
				"-print-... arguments require -E\n");
			exit(EX_USAGE);
		}
	}

	/*
	 * Ensure that there are some input files present.
	 */
	if(ac > optind) {
		ac -= optind;
		av += optind;
	} else {
		char *bin_name = a1c_basename(av[0]);
		fprintf(stderr, "%s: No input files specified. "
			"Try '%s -h' for more information\n",
			bin_name, bin_name);
		exit(1);
	}

	/*
	 * Make sure the skeleton directory is out there.
	 */
	if(skeletons_dir == NULL) {
		struct stat sb;
		skeletons_dir = DATADIR;
		if((av[-optind][0] == '.' || av[-optind][1] == '/')
		&& stat(skeletons_dir, &sb)) {
			/*
			 * The default skeletons directory does not exist,
			 * compute it from my file name:
			 * ./asn1c/asn1c -> ./skeletons
			 */
			char *p;
			size_t len;

			p = a1c_dirname(av[-optind]);

			len = strlen(p) + sizeof("/../skeletons");
			skeletons_dir = malloc(len);
			assert(skeletons_dir);
			snprintf(skeletons_dir, len, "%s/../skeletons", p);
			if(stat(skeletons_dir, &sb)) {
				fprintf(stderr,
					"WARNING: skeletons are neither in "
					"\"%s\" nor in \"%s\"!\n",
					DATADIR, skeletons_dir);
				if(warnings_as_errors)
					exit(EX_OSFILE);
			}
		}
	}

	/*
	 * Iterate over input files and parse each.
	 * All syntax trees from all files will be bundled together.
	 */
	for(i = 0; i < ac; i++) {
		asn1p_t *new_asn;

		new_asn = asn1p_parse_file(av[i], asn1_parser_flags);
		if(new_asn == NULL) {
			fprintf(stderr, "Cannot parse \"%s\"\n", av[i]);
			exit(EX_DATAERR);
		}

		/*
		 * Bundle the parsed tree with existing one.
		 */
		if(asn) {
			asn1p_module_t *mod;
			while((mod = TQ_REMOVE(&(new_asn->modules), mod_next)))
				TQ_ADD(&(asn->modules), mod, mod_next);
			asn1p_delete(new_asn);
		} else {
			asn = new_asn;
		}
	}

	/* These are mostly notes for the human readers */
	assert(asn);
	assert(skeletons_dir);

	/*
	 * Dump the parsed ASN.1 tree if -E specified and -F is NOT given.
	 */
	if(print_arg__print_out && !print_arg__fix_n_print) {
		if(asn1print(asn, asn1_printer_flags))
			exit(EX_SOFTWARE);
		return 0;
	}

	/*
	 * Read in the files from skeletons/standard-modules
	 */
	if(importStandardModules(asn, skeletons_dir)) {
		if(warnings_as_errors)
			exit(EX_DATAERR);
	}

	/*
	 * Process the ASN.1 specification: perform semantic checks,
	 * expand references, etc, etc.
	 * This function will emit necessary warnings and error messages.
	 */
	ret = asn1f_process(asn, asn1_fixer_flags,
		NULL /* default fprintf(stderr) */);
	switch(ret) {
	case 1:
		if(!warnings_as_errors)
			/* Fall through */
	case 0:
		break;			/* All clear */
	case -1:
		exit(EX_DATAERR);	/* Fatal failure */
	}

	/*
	 * Dump the parsed ASN.1 tree if -E specified and -F is given.
	 */
	if(print_arg__print_out && print_arg__fix_n_print) {
		if(asn1print(asn, asn1_printer_flags))
			exit(EX_SOFTWARE);
		return 0;
	}

	/*
	 * Compile the ASN.1 tree into a set of source files
	 * of another language.
	 */
	if(asn1_compile(asn, skeletons_dir, asn1_compiler_flags,
			ac + optind, optind - 1, av - optind)) {
		exit(EX_SOFTWARE);
	}

	return 0;
}

/*
 * Parse and import *.asn1 from skeletons/standard-modules
 */
static int
importStandardModules(asn1p_t *asn, const char *skeletons_dir) {
	asn1p_t *new_asn;
	asn1p_module_t *mod;
	const char *filename;
	char *fullname;
	char *target_dir;
	int target_dir_len;
	int len;
#ifdef	_WIN32
	intptr_t dir;
	struct _finddata_t c_file;
	char *pattern;
#else
	struct dirent *dp;
	DIR *dir;
#endif
	int ret = 0;

	/* Notes for the human reader */
	assert(asn);
	assert(skeletons_dir);

	/*
	 * Figure out the standard-modules directory.
	 */
	target_dir_len = strlen(skeletons_dir)
				+ sizeof("/standard-modules") - 1;
	target_dir = malloc(target_dir_len + 1);
	assert(target_dir);
	snprintf(target_dir, target_dir_len + 1, "%s/standard-modules",
		skeletons_dir);

#ifdef	_WIN32
	len = target_dir_len + sizeof("/*.asn1");
	pattern = malloc(len);
	assert(pattern);
	snprintf(pattern, len, "%s/*.asn1", target_dir);
	dir = _findfirst(pattern, &c_file);
	if(dir == -1L) {
#else
	dir = opendir(target_dir);
	if(!dir) {
#endif
		fprintf(stderr,
			"WARNING: Cannot find standard modules in %s\n",
			target_dir);
		return -1;
	}

#ifdef	_WIN32
	do {
		filename = c_file.name;
#else
	while((dp = readdir(dir))) {
		filename = dp->d_name;
#endif
		len = strlen(filename);
		if(len <= 5 || strcmp(filename + len - 5, ".asn1"))
			continue;
		len = target_dir_len + 1 + len + 1;
		fullname = malloc(len);
		if(!fullname) continue;	/* Just skip it, no big deal */
		snprintf(fullname, len, "%s/%s", target_dir, filename);
		filename = fullname;

		new_asn = asn1p_parse_file(filename, A1P_NOFLAGS);
		if(new_asn == NULL) {
			fprintf(stderr, "WARNING: Cannot parse standard module \"%s\"\n", filename);
			ret = -1;
			continue;
		}

		/* Import these modules and mark them as "standard" */
		while((mod = TQ_REMOVE(&(new_asn->modules), mod_next))) {
			mod->_tags |= MT_STANDARD_MODULE;
			TQ_ADD(&(asn->modules), mod, mod_next);
		}
		asn1p_delete(new_asn);

#ifdef	_WIN32
	} while(_findnext(dir, &c_file) == 0);
	_findclose(dir);
#else
		free(fullname);
	} /* while(readdir()) */
	closedir(dir);
#endif

	return ret;
}

/*
 * Print the usage screen and exit(EX_USAGE).
 */
static void
usage(const char *av0) {
	fprintf(stderr,
"ASN.1 Compiler, v" VERSION "\n" COPYRIGHT
"Usage: %s [options] file ...\n"
"Options:\n"
"  -E                    Run only the ASN.1 parser and print out the tree\n"
"  -F                    During -E operation, also perform tree fixing\n"
"\n"
"  -P                    Concatenate and print the compiled text\n"
"  -R                    Restrict output (tables only, no support code)\n"
"  -S <dir>              Directory with support (skeleton?) files\n"
"                        (Default is \"%s\")\n"
"  -X                    Generate and print the XML DTD\n"
"\n"

"  -Werror               Treat warnings as errors; abort if any warning\n"
"  -Wdebug-lexer         Enable verbose debugging output from lexer\n"
"  -Wdebug-fixer         --//-- semantics processor\n"
"  -Wdebug-compiler      --//-- compiler\n"
"\n"

"  -fbless-SIZE          Allow SIZE() constraint for INTEGER etc (non-std.)\n"
"  -fcompound-names      Disambiguate C's struct NAME's inside top-level types\n"
"  -findirect-choice     Compile members of CHOICE as indirect pointers\n"
"  -fincludes-quoted     Generate #includes in \"double\" instead of <angle> quotes\n"
"  -fknown-extern-type=<name>    Pretend the specified type is known\n"
"  -fline-refs           Include ASN.1 module's line numbers in comments\n"
"  -fno-constraints      Do not generate constraint checking code\n"
"  -fno-include-deps     Do not generate courtesy #includes for dependencies\n"
"  -funnamed-unions      Enable unnamed unions in structures\n"
"  -fwide-types          Use INTEGER_t instead of \"long\" by default, etc.\n"
"\n"

"  -gen-PER              Generate PER support code\n"
"  -pdu={all|auto|Type}  Generate PDU table (discover PDUs automatically)\n"
"\n"

"  -print-class-matrix   Print out the collected object class matrix (debug)\n"
"  -print-constraints    Explain subtype constraints (debug)\n"
"  -print-lines          Generate \"-- #line\" comments in -E output\n"

	,
	a1c_basename(av0), DATADIR
	);
	exit(EX_USAGE);
}

